File: tnf.py
VAR, REF, PAIR, CONST = 1,2,3,4

# unnamed variable
def var() : return [None]

# named variable
def nvar(name): return [None,name]

# return type of data object
def type_of(x) :
  if isinstance(x,list) :
    if x[0] is None : return VAR
    return REF
  elif isinstance(x,tuple) :
    return PAIR
  else:
    return CONST

# finding the reference of a variable
def deref(x) :
  while True :
    t=type_of(x)
    if t==REF:
      x=x[0]
    else:
     return x,t

# return term with all vars in it dereferenced
def iron(a,size=200) :
  def iron0(a, k):
    if k <= 0: return ".?."
    x, t = deref(a)
    if t == PAIR:
      u, v = x
      k -= 1
      return iron0(u, k), iron0(v, k)
    else:
      return x
  return iron0(a,size)

def to_pterm(x) :
  x,t=deref(x)
  if VAR==t : return x
  elif CONST==t : return x
  else :
    assert PAIR==t
    a, b = x
    args=[None,to_pterm(a)]
    while True:
      b, t = deref(b)
      if PAIR == t :
        a,b=b
        args.append(to_pterm(a))
      else :
        args[0]=to_pterm(b)
        return list(args)

# converts a "."/2 lists to actual lists
def dot2list(a):
  def is_dot_list(a):
    return isinstance(a,list) and len(a)>0 and '.'==a[0]
  def walk(x) :
    if is_dot_list(x) :
      args=[]
      while is_dot_list(x) :
        e=dot2list(x[1])
        args.append(e)
        y=x
        x=x[2]
      if x=='[]' :
        return args
      else :
        y=(y[0],dot2list(y[1]),dot2list(y[2]))
        args.append(y)
        return args
    elif VAR==type_of(x) :
      return x
    elif isinstance(x,tuple) or isinstance(x,list) :
      return tuple(map(dot2list,x))
    else :
      return x
  return walk(a)

def pp(x) : return dot2list(to_pterm(x))

# turn term into more readable string
def pt(a,size=200) :
  def pt0(a, k):
    if k <= 0: return ".?."
    x, t = deref(a)
    if t == PAIR:
      u, v = x
      k -= 1
      return f'({pt0(u, k)}=>{pt0(v, k)})'
    elif t == VAR:
      if len(x) == 2:
        return str(x[1])
      else:
        return "_" + str(id(x))
    else:
      return str(x)
  return pt0(a,size)


# creates actual variables from code skeletons
def activate(instr,vars,vids) :
  newinstr=[]
  #l=len(vars)
  for c in instr :
    if isinstance(c,list): # a variable
      n=c[0]
      if n<len(vars) :
        d=vars[n]
      else:
        d=nvar(c[1])
        #print('D=',d,id(d),type(vids))
        vids.add(id(d))
        vars.append(d)
    else: # const
      d=c
    newinstr.append(d)
  return newinstr

# undo bindings on the trail
def unwind(trail,ttop) :
  #print('UNWIND',pt(trail[-1]),':',len(trail),'-',ttop)
  for _ in range(len(trail)-ttop):
    v = trail.pop()
    v[0] = None

# unify two terms
def unify(x,y,trail,vids) :
  ustack=[]
  ustack.append(y)
  ustack.append(x)
  while ustack :
    x1,t1=deref(ustack.pop())
    x2,t2=deref(ustack.pop())
    if x1 is x2:
      pass
    elif VAR is t1 :
      x1[0]=x2
      if id(x1) not in vids:
        trail.append(x1)
    elif VAR is t2:
      x2[0]=x1
      if id(x2) not in vids:
        trail.append(x2)
    elif t1!=t2 :
      return False
    elif t1 is CONST:
      if x1!=x2 :
        return False
    else :
      assert t1 == PAIR and len(x1)==2
      assert t2 == PAIR and len(x2)==2
      a1,b1=x1
      a2,b2=x2   
      ustack.append(b2)
      ustack.append(b1)
      ustack.append(a2)
      ustack.append(a1)
  return True

# builds goal assuming goal(X):- ... is
# a clause in the source program
def get_goal() :
  answer= nvar('Answer')
  return  (answer,('true','goal')),answer

FAIL, DONE, DO, UNDO = 0, 1, 2, 3
opnames=["FAIL", "DONE", "DO", "UNDO"]

# inner loop of at most len(code) steps
def step(G,i, code,trail):
    ttop=len(trail)
    while i<len(code):
      unwind(trail, ttop)
      clause=code[i]
      i+=1 #next clause
      vars,vids = [],set()
      #print('@@@ CLAUSE:',clause)
      for instr in clause :
        #print("INSTR:",instr)
        c=activate(instr,vars,vids)
        op=c[0]
        #print("!!!",c)

        if op=='u' :
          old,oldt=deref(c[3])
          if VAR==oldt :
            old[0]=c[1],c[2]
            continue
          else :
            #print("OLD:",old)
            assert PAIR==oldt
            if not (unify(c[1],old[0],trail,vids) and \
                    unify(c[2],old[1],trail,vids)) : break

        elif op=='b':
          old,oldt=deref(c[3])
          #assert VAR==oldt
          old[0]=c[1],c[2]
          continue

        elif op=='d':
          assert VAR==type_of(c[1])
          c[1][0]=G
          continue

        else: #p(..)
          NewG,tg=deref(c[1])
          if NewG=='true' and CONST == tg :
            return (DONE,G,ttop,i)
          else:
            assert PAIR==tg
            return (DO,(NewG,G),ttop,i)
    return (FAIL,None,ttop,i)


def ensure_undo(G,todo,ttop,i,l) :
    #if i==l-1: print('!!!',i)
    instr = (UNDO, G, ttop, i)
    if todo :
       (op, G1, ttop1, i1)=todo[-1]
       if (op,i1,G)==(UNDO,i,G1) and ttop>=ttop1:
         return
    todo.append(instr)

def show_todo(todo) :
  k=len(todo)
  while k>0:
    k-=1
    (op,G,ttop,i)=todo[k]
    print("v",opnames[op],ttop,i)
    if(op==UNDO) : print(pp(todo[k][1]))
  print("-----------\n")



# code interpreter
def interp(code) :

  # trampoline, ensures LCO, no recursion etc.

  goal,answer=get_goal()
  l=len(code)
  todo=[(DO,(goal,None),0,l)]
  ctr,maxctr=0, 200000000000
  max_trail, max_todo = 0, 0
  trail=[]
  while ctr<maxctr and todo :

    #show_todo(todo)

    ctr+=1
    max_trail=max(max_trail,len(trail))
    max_todo = max(max_todo, len(todo))
    op,G,ttop,i=todo.pop()
    #opn=opnames[op]
    #print('\n',ctr,'!!!','op=',opn,'ttop=',len(trail),'todo=',len(todo))
    #print('--->',pp(instr[1]),instr[2],'i=',instr[3])

    if DO==op :
      (NewG,OldG)=G
      if i<l : ensure_undo(OldG,todo,ttop,i,l);
      todo.append(step(NewG, 0, code, trail))

    elif DONE==op:
      yield iron(answer), max_trail, max_todo
      if i<l: ensure_undo(G,todo,ttop,i,l)

    elif UNDO==op :
      unwind(trail, ttop)
      if i<l: todo.append(step(G,i, code,trail))

    else: # FAIL == op:
      pass

  if ctr==maxctr: print('*** TOO MANY STEPS? ',ctr)
  print('max_trail=',max_trail,'max_todo=',max_todo)

# tests
def go() :
  code=tload()
  ctr=0
  max_ctr=10000000000
  for a,mtr,mtd in interp(code):
    ctr+=1
    print('ANSWER=',ctr,'trail=',mtr,'todo=',mtd,'-->',pp(a))
    if ctr>=max_ctr :
      print('TOO MANY ANSWERS?')
      break
  print('TOTAL ANSWERS:',ctr)

# load from an "assembler" code file
def tload(fname='out/tnf_asm.txt'):
  def to_const(x) :
    try :
      return int(x)
    except ValueError:
      try:
        return float(x)
      except ValueError:
        return x
  with open(fname,'r') as f: txt=f.read()
  ls = txt.split('\n')
  code = []
  vars=dict()
  cs=[]
  for l in ls:
    xs=l.split(' ')
    if len(xs)<=1 :continue # skip extra new lines
    for j,x in enumerate(xs) :
      if x[0].isupper():
        if x in vars:
          v=vars[x]
        else:
          v=[len(vars),x]
          vars[x]=v
        xs[j]=v
      elif x[0]=="_":
        v = [len(vars), x]
        vars[x] = v
        xs[j] = v
      elif x=='[|]':
        xs[j]='.'
      else :
        xs[j] = to_const(x)
    cs.append(xs)
    if xs[0]=='p' :
      code.append(cs)
      vars=dict() # new vars scoped over next clause
      cs=[]
  return code

# tests unification
def utest() :
  X=var()
  Y=var()
  a=(X,'a')
  b=('b',Y)
  print('AB',a,type_of(a),b,type_of(b))
  trail=[]
  r=unify(a,b,trail,set())
  print(r)
  print('FINAL AB',a,type_of(a),b,type_of(b))
  print('TR',trail)
  print("PT",pt((a,(nvar("X"),var()))))

# tests conversion to Horn form
def ptest():
  A = nvar('A')
  B = nvar('B')
  print(to_pterm((((A, (A, 'b')), 'a'), (B, 'c'))))
  print(to_pterm(((1, 'g'), (2, (3, 'f')))))

# also works with gc disabled
#import gc
#gc.disable()
go()
#utest()

File: julia/tnf.jl
VAR, REF, PAIR, CONST = 1, 2, 3, 4

mutable struct Var
    val::Any
    age::Int
    name::String
end

# new variables have empty as their value
empty="#"

# creates a new variable
function ivar(age, name)
    Var(empty, age, name)
end

# returns the type of our terms
function type_of(x)
    if x isa Var
        if x.val == empty
            VAR
        else
            REF
        end
    elseif x isa Pair
        PAIR
    else
        CONST
    end
end

# binds a variable to a value, if born before age
function bindVar(v,t,trail,ctr)
    v.val = t
    if v.age<ctr
        push!(trail,v)
    end
end

# dereferences a variable over a chain of value slots
function deref(x)
    #println("DEREF<<<:",x,':',type_of(x))
    while true
        t = type_of(x)
        if t == REF
            x = x.val
        else
            #println("DEREF>>>:",x)
            @assert(x!=empty,"!!!!NOT EMPTY")
            return x, t
        end
    end
end

# unifies two terms, places them on thrail if born before ctr
function unify(x,y,trail,ctr)
    ustack=[]
    push!(ustack,y)
    push!(ustack,x)
    while(!isempty(ustack))
        x1,t1=deref(pop!(ustack))
        x2,t2=deref(pop!(ustack))
        #println("### UNIFY ");pt(x1);pt(x2);println()
        if x1===x2
            continue
        elseif VAR==t1 && VAR==t2
          if x1.age>x2.age
              bindVar(x1,x2,trail,ctr)
          else
              bindVar(x2,x1,trail,ctr)
          end
        elseif VAR==t1
            bindVar(x1,x2,trail,ctr)
        elseif VAR==t2
            bindVar(x2,x1,trail,ctr)
        elseif t1!=t2
            return false
        elseif t1==CONST
            if x1!=x2
                return false
            end
        else
            @assert(PAIR==t1,"PAIR expected")
            a1,b1=x1
            a2,b2=x2
            push!(ustack,b2)
            push!(ustack,b1)
            push!(ustack,a2)
            push!(ustack,a1)
        end
    end
    return true
end

# undoes binding by reseting vars on the trail
function unwind(trail,ttop)
  l=length(trail)
  #println("ENTER UNW ",l,',',ttop)
  for _ in ttop+1:l
      v=pop!(trail)
      #println("UNW",pt(v))
      v.val=empty
  end
end



fuel=1000000000

# basic instruction  processing step
function step_(G,i,code,trail,varctr)
    ttop=length(trail)
    l=length(code)+1
    fail=(FAIL,nothing,ttop,i)
    global fuel;fuel-=1;if fuel<=0;return fail;end
    while i<l
        unwind(trail,ttop)
        clause,k=code[i]
        vs=[]
        i+=1
        #println("\n!!! ENTERING step: i=",i-1)
        for templ in clause
            (op,c1,c2,c3)=instr=activate(templ,k,vs)
            if op=='u'
                old,oldt=deref(c3)
                if VAR==oldt
                    old.val=Pair(c1,c2)
                    continue
                else
                    @assert(PAIR==oldt,pt(c3))
                    ok= unify(c1,old.first,trail,varctr.val)
                    if !ok
                        break
                    end
                    ok= unify(c2,old.second,trail,varctr.val)
                    if !ok
                        break
                    end
                end
                continue
            elseif op=='b'
                old,oldt=deref(c3)
                old.val=Pair(c1,c2)
                continue
            elseif op=='d'
                c1.val=G
                continue
            else
                @assert(op=='p',"EXPECTED: p")
                varctr.val+=k
                NewG,tg=deref(c1)

                if NewG=="true" && CONST == tg
                    return (DONE,G,ttop,i)
                else
                    @assert(PAIR==tg,"!!!PAIR")
                    return (DO,(NewG,G),ttop,i)
                end
            end
        end
    end
    fail
end

# trampoline state machine states
FAIL,DONE,DO,UNDO=1,2,3,4

opn=["FAIL","DONE","DO","UNDO"]

function ensure_undo(G,todo,ttop,i,l)
    instr = (UNDO, G, ttop, i)
    push!(todo,instr)
end

# goal for entry point convention
# assuming execution starts from predicate goal(ANSWER)
# and when found, answer is printed out
function get_goal()
    answer=ivar(0,"ANSWER")
    Pair(answer,Pair("true","goal"))
end

function interp()
  global fuel;fuel-=1;if fuel<=0;return;end
  code=tload()
  l=length(code)+1
  todo=[];trail=[]
  goal=get_goal()
  answer=goal.first
  push!(todo,(DO,(goal,nothing),0,l+1))

  varctr=ctr(1)
  while(!isempty(todo))
      action=pop!(todo)
      (op,G,ttop,i)=action

      if DO==op # advance forward oin AND branch
          NewG,OldG=G
          if i<l
              ensure_undo(OldG,todo,ttop,i,l)
          end
          r=step_(NewG,1,code,trail,varctr)
          push!(todo,r)
      elseif DONE==op # show ANSWER
          println(">>>ANSWER:",pt(answer))
          if i<l
              ensure_undo(G,todo,ttop,i,l)
          end
      elseif UNDO==op # explore next OR alternative
          unwind(trail,ttop)
          if i<l
              push!(todo,step_(G,i,code,trail,varctr))
          end
      else # FAIL - done with finding matching clauses
      end
  end
end

# IO

# more human readable string representation
function pt(z, size = 2000)
    if z==nothing;return z;end
    function pt0(a, k)
        if k <= 0
            return "<?>"
        else
            x, t = deref(a)
            if t == PAIR
                u, v = x
                k -= 1
                s1=pt0(u, k)
                s2=pt0(v, k)
                return string("(", s1, "->", s2, ")")
            elseif t == VAR
                #return string("_", string(objectid(x)))
                return string(x.name,"",x.age)
            else
                @assert(CONST==t,"should be CONST!")
                return string(x)
            end
        end
    end
    return pt0(z, size)
end

# pair made of a dictionary and a list
mutable struct ST
    d::Any
    a::Any
end

# creats empty dict+array pair
function new_st()
    ST(Dict(), [])
end

# add clause var inless seen already
function add_cvar(w,st,ctr)
    if haskey(st.d, w)
        st.d[w]
    else
        ctr.val+=1
        push!(st.a, w)
        l = length(st.a)
        v = ivar(l,w)
        st.d[w] = ivar(l, w)
        v
    end
end

# vars are uppercvas othes are constants
function vc(w,st,ctr)::Any
    if isuppercase(w[1])
        add_cvar(w,st,ctr)
    else
        w
    end
end

# extract instruction template from tokes on a line
function to_instr(toks,st,ctr)
    op = toks[1]
    @assert(op!=nothing,"OP!!!")
    if (op == "d") || (op == "p")
        instr = (op[1], vc(toks[2],st,ctr), nothing, nothing)
    else
        instr = (op[1], vc(toks[2],st,ctr),
        vc(toks[3],st,ctr), vc(toks[4],st,ctr))
    end
end

# mutable counter
mutable struct ctr
    val
end

# creates new variable from templat unless arleady in vs
function nv(t,k,vs)
    if t isa Var
        if(t.age<=length(vs))
            return vs[t.age]
        else
          v=ivar(t.age+k,t.name)
          push!(vs,v)
          return v
      end
    else
        t
    end
end

# activates variables for instruction template
function activate(ts,k,vs)
   ts[1],nv(ts[2],k,vs),nv(ts[3],k,vs),nv(ts[4],k,vs)
end

function tload(fname = "../out/tnf_asm.txt")
    lines = readlines(fname)
    st=new_st()
    code=[]
    cls=[]
    count=ctr(0)
    for line in lines
        if length(line) < 2
            continue
        end
        #println("LINE:",line)
        toks = split(line)
        #for t in toks;println("TOK=",String(t));end;println()
        push!(cls,to_instr(toks,st,count))
        if toks[1]=="p"
            push!(code,(cls,count.val))
            st=new_st()
            cls=[]
            count=ctr(0)
        end
    end

    for i in 1:length(code)
        c,count=code[i]
        vs=[]
        for x in c
            #println('X',deref(x[1]))
            #pti(x)
            a=activate(x,100,vs)

        end
        #println("count: i=",i," var count=",count) #;ptrail(vs)
    end

    code
end

# show trail
function ptrail(xs)
    println("TRAIL:")
       for x in xs
           println(pt(x))
       end
       println("")
   end

# show todo stack
function ptodo(xs)
       println("TODO:")
          for x in xs
              pta(x)
          end
          println("")
end

# show instruction
function pti(instr)
    s=" , "
    op,x1,x2,x3=instr
    println("<",op,">::",pt(x1),s,pt(x2),s,pt(x3))
end

# show state machine action
function pta(action)
    s=" , "
    act,G,ttop,i=action
    if act==DO
        println(opn[act],": ",pt(G[1]),s,pt(G[2]),s,ttop,i)
    else
        println(opn[act],": ",pt(G),s,s,ttop,i)
    end
end

########## tests ##########
function test()
    println("TEST")
    k=10
    trail=[]
    v=ivar(1,"V")
    u=ivar(2,"U")
    x=ivar(3,"X")
    y=ivar(4,"Y")
    bindVar(x,v,trail,k)
    bindVar(y,u,trail,k)
    println("DEREF: ",pt(x))
    p1=Pair(x,"a")
    p2=Pair("b",y)

    println("UNIF BEFORE: ",pt(p1)," ==== ",pt(p2))
    unify(p1,p2,trail,k)
    println("UNIF AFTER:  ",pt(p1)," ==== ",pt(p2))
    ptrail(trail)
    println(length(trail))
    unwind(trail,0)
    println("UNWINDED")
    ptrail(trail)

end

#clearconsole()

#test()
interp()
